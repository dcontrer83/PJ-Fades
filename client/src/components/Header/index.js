import React from 'react'

import Container from 'react-bootstrap/Container';
import Navbar from 'react-bootstrap/Navbar';
import Nav from 'react-bootstrap/Nav';

import logo from '../../images/logo/logo.png'

import Auth from '../../utils/auth';

function Header(props) {
  const logout = (event) => {
    event.preventDefault();
    Auth.logout();
  }

  const style = {
    container: {
      margin: "0",
      maxWidth: "100%",
      fontColor: "white",
    },
    image: {
      width: "92px",
      height: "75px"
    },
    nav: {
      width: "60%",
      marginRight: "10%",
      fontSize: "20px"
    }
  }

  return (
    <Navbar data-testid="navbar-element" className="fixed-top" bg="light" expand="sm">
      <Container data-testid="container-element" style={style.container}>
        <Navbar.Brand data-testid="brand-element" href="/">𝐏𝐉-𝐅𝐀𝐃𝐄𝐒</Navbar.Brand>
        <img data-testid="img-element" src={logo} alt="logo" style={style.image}></img>
        <Navbar.Toggle aria-controls="navbar" />
        <Container fluid="lg" style={style.container}>
          <Navbar.Collapse data-testid="collapse-element" className="justify-content-end" id="navbar">
            <Nav data-testid="nav-element" className="justify-content-around" style={style.nav}>
              <Nav.Link data-testid="home-element" href="/">Home</Nav.Link>
              <Nav.Link data-testid="gallery-element" href="/gallery">Gallery</Nav.Link>
              <Nav.Link data-testid="contact-element" href="/contact">Contact</Nav.Link>
              <Nav.Link data-testid="reservation-element" href="/reservation">Reservation</Nav.Link>
              {Auth.loggedIn() ? (
                // <button className="btn btn-lg btn-light m-2" onClick={logout}>
                // Logout
                // </button>
                <>
                  <Nav.Link data-testid="profile-element" href="/profile">Profile</Nav.Link>
                  <Nav.Link data-testid="logout-element" href="/" onClick={logout}>Logout</Nav.Link>
                </>
              ) : (
                <Nav.Link data-testid="login-element" href="/login">Login</Nav.Link>
              )}
              {/* <Nav.Link href="/login">Login</Nav.Link> */}
            </Nav>
          </Navbar.Collapse>
        </Container>
      </Container>
    </Navbar>
  )
}
export default Header;