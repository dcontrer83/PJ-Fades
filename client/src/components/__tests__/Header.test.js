import React from 'react';
import { createRoot } from 'react-dom/client';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import Header from '../Header';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import Auth from '../../utils/auth';

let container;
let root;
let navbarElement;
let containerElement;
let brandElement;
let imgElement;
let collapseElement;
let navElement;
let homeElement;
let galleryElement;
let contactElement;
let reservationElement;
let loginElement;
let profileElement;
let logoutElement;

globalThis.IS_REACT_ACT_ENVIRONMENT = true;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    act(() => {
        root = createRoot(container);
    })
});

afterEach(() => {
    act(() => {
        root.unmount();
    })
    container.remove();
    container = null;
});

describe('Header', () => {
    it('should match snapshot', () => {
        act(() => {
            root.render(<Header />);
        });
        expect(pretty(container.innerHTML)).toMatchSnapshot();
    });
});

describe('Navbar Element', () => {
    it('should be in the document', () => {
        getNavbarElement();
        expect(navbarElement).toBeInTheDocument();
    });
    it('should have specific className', () => {
        getNavbarElement();
        expect(navbarElement).toHaveClass('fixed-top navbar navbar-expand-sm navbar-light bg-light');
    })
});

describe('Container Element', () => {
    it('should be in the document', () => {
        getContainerElement();
        expect(containerElement).toBeInTheDocument();
    });
    it('should have specific style', () => {
        getContainerElement();
        expect(containerElement).toHaveStyle('margin: 0; maxWidth: 100%;');
    });
});

describe('Brand Element', () => {
    it('should be in the document', () => {
        getBrandElement();
        expect(brandElement).toBeInTheDocument();
    });
    it('should redirect to / when clicked', () => {
        getBrandElement();
        expect(brandElement.href).toBe('http://localhost/');
    });
});

describe('Img Element', () => {
    it('should be in the document', () => {
        getImgElement();
        expect(imgElement).toBeInTheDocument();
    });
    it('should have proper src and alt attributes', () => {
        getImgElement();
        expect(imgElement).toHaveAttribute('src', 'logo.png');
        expect(imgElement).toHaveAttribute('alt', 'logo');
    });
    it('should have specific style', () => {
        getImgElement();
        expect(imgElement).toHaveStyle('width: 92px; height: 75px');
    });
});

describe('Collapse Element', () => {
    it('should be in the document', () => {
        getCollapseElement();
        expect(collapseElement).toBeInTheDocument();
    });
    it('should have specific className', () => {
        getCollapseElement();
        expect(collapseElement).toHaveClass('justify-content-end');
    });
});

function getNavElement() {
    render(<Header />);
    navElement = screen.getByTestId('nav-element');
};

describe('Nav Element', () => {
    it('should be in the document', () => {
        getNavElement();
        expect(navElement).toBeInTheDocument();
    });
    it('should have specific className', () => {
        getNavElement();
        expect(navElement).toHaveClass('justify-content-around');
    });
    it('should have specific style', () => {
        getNavElement();
        expect(navElement).toHaveStyle('width: 60%; marginRight: 10%; fontSize: 20px');
    });
});


function getNavLinks() {
    render(<Header />);
    homeElement = screen.getByTestId('home-element');
    galleryElement = screen.getByTestId('gallery-element');
    contactElement = screen.getByTestId('contact-element');
    reservationElement = screen.getByTestId('reservation-element');
    loginElement = screen.getByTestId('login-element');
}

describe('Nav Link Elements', () => {
    it('should be in the document', () => {
        getNavLinks();
        expect(homeElement).toBeInTheDocument();
        expect(galleryElement).toBeInTheDocument();
        expect(contactElement).toBeInTheDocument();
        expect(reservationElement).toBeInTheDocument();
        expect(loginElement).toBeInTheDocument();
    });
    it('should redirect to proper link', () => {
        getNavLinks();
        expect(homeElement.href).toBe('http://localhost/');
        expect(galleryElement.href).toBe('http://localhost/gallery');
        expect(contactElement.href).toBe('http://localhost/contact');
        expect(reservationElement.href).toBe('http://localhost/reservation');
        expect(loginElement.href).toBe('http://localhost/login');
    });
});

function getLoggedInElements() {
    render(<Header />);
    profileElement = screen.getByTestId('profile-element');
    logoutElement = screen.getByTestId('logout-element');
}

describe('Profile and Logout link when logged in', () => {
    it('should be in the document', () => {
        jest.spyOn(Auth, 'loggedIn').mockReturnValue(true);
        getLoggedInElements();
        expect(profileElement).toBeInTheDocument();
        expect(logoutElement).toBeInTheDocument();
    });
    it('should redirect to proper link', () => {
        jest.spyOn(Auth, 'loggedIn').mockReturnValue(true);
        getLoggedInElements();
        expect(profileElement.href).toBe('http://localhost/profile');
        expect(logoutElement.href).toBe('http://localhost/');
    });
});

function getNavbarElement() {
    render(<Header />);
    navbarElement = screen.getByTestId('navbar-element');
}

function getContainerElement() {
    render(<Header />);
    containerElement = screen.getByTestId('container-element');
}

function getBrandElement() {
    render(<Header />);
    brandElement = screen.getByTestId('brand-element');
}

function getImgElement() {
    render(<Header />);
    imgElement = screen.getByTestId('img-element');
}

function getCollapseElement() {
    render(<Header />);
    collapseElement = screen.getByTestId('collapse-element');
}