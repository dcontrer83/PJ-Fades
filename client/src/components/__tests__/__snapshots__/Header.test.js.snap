// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot 1`] = `
"<nav data-testid=\\"navbar-element\\" class=\\"fixed-top navbar navbar-expand-sm navbar-light bg-light\\">
  <div data-testid=\\"container-element\\" style=\\"margin: 0px; max-width: 100%;\\" class=\\"container\\"><a data-testid=\\"brand-element\\" href=\\"/\\" class=\\"navbar-brand\\">𝐏𝐉-𝐅𝐀𝐃𝐄𝐒</a><img data-testid=\\"img-element\\" src=\\"logo.png\\" alt=\\"logo\\" style=\\"width: 92px; height: 75px;\\"><button aria-controls=\\"navbar\\" type=\\"button\\" aria-label=\\"Toggle navigation\\" class=\\"navbar-toggler collapsed\\"><span class=\\"navbar-toggler-icon\\"></span></button>
    <div style=\\"margin: 0px; max-width: 100%;\\" class=\\"container-lg\\">
      <div class=\\"justify-content-end navbar-collapse collapse\\" data-testid=\\"collapse-element\\" id=\\"navbar\\">
        <div class=\\"justify-content-around navbar-nav\\" data-testid=\\"nav-element\\" style=\\"width: 60%; margin-right: 10%; font-size: 20px;\\"><a data-testid=\\"home-element\\" href=\\"/\\" data-rr-ui-event-key=\\"/\\" class=\\"nav-link\\">Home</a><a data-testid=\\"gallery-element\\" href=\\"/gallery\\" data-rr-ui-event-key=\\"/gallery\\" class=\\"nav-link\\">Gallery</a><a data-testid=\\"contact-element\\" href=\\"/contact\\" data-rr-ui-event-key=\\"/contact\\" class=\\"nav-link\\">Contact</a><a data-testid=\\"reservation-element\\" href=\\"/reservation\\" data-rr-ui-event-key=\\"/reservation\\" class=\\"nav-link\\">Reservation</a><a data-testid=\\"login-element\\" href=\\"/login\\" data-rr-ui-event-key=\\"/login\\" class=\\"nav-link\\">Login</a></div>
      </div>
    </div>
  </div>
</nav>"
`;
